### DISCUSSION TITLE
Question: about developing rlr

### DISCUSSION THREAD

**Discussion by lf94** [epilys/rlr#2] [General]
Hey! While looking around for an on-screen ruler, I found nothing suitable for my system. A quick GitHub search brought me here. I was very curious to see how GTK was used and I have to say this may be some of the cleanest Rust-GTK code I've seen. I really want to ask: how did you figure out what to use, and how everything was used together? Are you very familiar with GTK?

Thank you for a great tool :) I've going to use it for making sure UI components are aligned

**Comment by epilys** [comment]
I'm very glad you found this little tool useful! And thank you so much for the compliments, it honestly means a lot to me.

I made this completely "from scratch". First I saw a `gtk-rs` example of how to set up an empty app and show a window when it launches, then looked at the available gtk widgets to see how I could draw graphics. I found the `DrawingArea` widget and then started reading the `cairo-rs` API :)

`gtk-rs`, `gio-rs` etc has everything documented since (I think) it's automatically generated from the C API. So you can look up stuff on the rust docs very fast, see what each type or trait does, and if you can't figure out something you can look it up online for GTK in C and try to do something analogous in Rust. 

`gtk-rs` has also lots of great examples in its git repository, which you can check out. (They weren't useful for `rlr` but they are useful in general).

Finally, some time ago I made this template/scaffold rust/gtk project based on the way I was currently structuring applications with `gtk-rs`. I made it for someone who was also interested in gtk/rust. It does nothing and has no content, but has code to setup an `Application`, `Window`, keyboard actions (accelerators), system menu and an about popup window. You can find it here https://github.com/epilys/fantastic-rust-gtk-adventure

After `rlr` I started this more complex gtk app (a font editor) but the complexity/verbosity of making data-focused apps in gtk kinda slowed down development and I don't know when/if I will pick it up again. But you can run it and play with fonts if you want. https://github.com/epilys/gerb

**Comment by lf94** [comment]
Thank you very, very much for the detailed response!

So it was quite some effort then to get ready to do things. This was pretty much my experience also.

Are you using anything else for UIs these days if you find gtk-rs too verbose/complex? (I'm familiar with the ecosystem)

**Comment by epilys** [comment]
If you want the full UI experience, I think the only choice is using gtk or cocoa (Apple/macos but also gnustep's implementation). Nothing else is as mature feature-wise except maybe tcl/tk which is "old".

See this article here about the complexity of UI frameworks: https://www.cmyr.net/blog/gui-framework-ingredients.html written by @cmyr who maintains a GUI framework in Rust, [Druid](https://github.com/linebender/druid)

If you want something quick and simple, I recommend the [tkinter python3 module](https://docs.python.org/3/library/tkinter.html#module-tkinter) which uses tcl/tk. Here's a very smol single-purpose app I wrote in tkinter https://github.com/epilys/scrcpy-input.py which was incredibly fast and easy to write without prior experience.

PS.

> So it was quite some effort then to get ready to do things

Looking at the commit history, it took 5 days :D
